File: yams/cpp/src/yams/mechanisms/positional/Elevator.cpp
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

#include "yams/mechanisms/positional/Elevator.hpp"

#include <wpi/framework/RobotBase.hpp>
#include <wpi/math/geometry/Rotation3d.hpp>
#include <wpi/math/geometry/Translation3d.hpp>
#include <wpi/simulation/BatterySim.hpp>
#include <wpi/simulation/RoboRioSim.hpp>
#include <wpi/smartdashboard/Mechanism2d.hpp>
#include <wpi/smartdashboard/MechanismLigament2d.hpp>
#include <wpi/smartdashboard/MechanismRoot2d.hpp>
#include <wpi/smartdashboard/SmartDashboard.hpp>
#include <wpi/util/Color.hpp>
#include <wpi/util/Color8Bit.hpp>
#include <wpi/commands2/Commands.hpp>
#include <wpi/units/length.hpp>
#include <wpi/units/time.hpp>

#include <array>
#include <cmath>
#include <memory>
#include <numbers>
#include <string>

#include "yams/exceptions.hpp"
#include "yams/gearing/MechanismGearing.hpp"
#include "yams/motorcontrollers/simulation/ElevatorSimSupplier.hpp"

namespace yams::mechanisms::positional {

// ---- Constructor ------------------------------------------------------------

Elevator::Elevator(config::ElevatorConfig* config, motorcontrollers::SmartMotorController* smc)
    : SmartPositionalMechanism() {
  m_elevatorConfig = config;
  m_smc = smc;
  m_subsystem = m_smc->GetConfig().GetSubsystem();

  if (!m_elevatorConfig->GetTelemetryName().empty()) {
    m_name = m_elevatorConfig->GetTelemetryName();
  }

  // Apply soft limits to the motor controller when configured.
  if (auto minH = m_elevatorConfig->GetMinHeight()) {
    m_smc->SetMeasurementLowerLimit(*minH);
  }
  if (auto maxH = m_elevatorConfig->GetMaxHeight()) {
    m_smc->SetMeasurementUpperLimit(*maxH);
  }

  // Seed the encoder from the configured starting position.
  if (auto startPos = m_smc->GetConfig().GetStartingPosition()) {
    m_smc->SetEncoderPosition(m_smc->GetConfig().ConvertFromMechanism(*startPos));
  }

  if (wpi::RobotBase::IsSimulation()) {
    // Configuration checks.
    if (!m_elevatorConfig->GetCarriageMass().has_value()) {
      throw exceptions::ElevatorConfigurationException("Mass is not configured!",
                                                       "Cannot create simulator",
                                                       "WithCarriageMass(wpi::units::kilogram_t)");
    }
    if (!m_elevatorConfig->GetMinHeight().has_value()) {
      throw exceptions::ElevatorConfigurationException("Minimum height is not configured!",
                                                       "Cannot create simulator",
                                                       "WithMinimumHeight(wpi::units::meter_t)");
    }
    if (!m_elevatorConfig->GetMaxHeight().has_value()) {
      throw exceptions::ElevatorConfigurationException("Maximum height is not configured!",
                                                       "Cannot create simulator",
                                                       "WithMaximumHeight(wpi::units::meter_t)");
    }
    if (!m_smc->GetConfig().GetStartingPosition().has_value()) {
      throw exceptions::ElevatorConfigurationException("Starting position is not configured!",
                                                       "Cannot create simulator",
                                                       "smc.WithStartingPosition(wpi::units::meter_t)");
    }
    if (!m_smc->GetConfig().GetMechanismCircumference().has_value()) {
      throw exceptions::ElevatorConfigurationException(
          "Mechanism circumference is not configured!", "Cannot create simulator",
          "SMC.WithMechanismCircumference(wpi::units::meter_t)");
    }

    wpi::math::DCMotor dcMotor = m_smc->GetDCMotor();
    auto& gearingOpt = m_smc->GetConfig().GetMotorGearing();
    gearing::MechanismGearing gearing = gearingOpt.value_or(gearing::MechanismGearing::kOne);

    bool simulateGravity = !m_elevatorConfig->IsHorizontal();
    wpi::units::meter_t circumference = m_smc->GetConfig().GetMechanismCircumference().value();

    wpi::units::meter_t startH =
        m_smc->GetConfig().ConvertFromMechanism(*m_smc->GetConfig().GetStartingPosition());

    m_elevatorSim.emplace(
        dcMotor, gearing.GetMechanismToRotorRatio(), m_elevatorConfig->GetCarriageMass().value(),
        circumference / (2.0 * std::numbers::pi), m_elevatorConfig->GetMinHeight().value(),
        m_elevatorConfig->GetMaxHeight().value(), simulateGravity, startH,
        std::array<double, 2>{0.01 / 4096.0, 0.01 / 4096.0});

    wpi::units::second_t period = m_smc->GetConfig().GetClosedLoopControlPeriod().value_or(20_ms);
    m_smc->SetSimSupplier(std::make_shared<yams::motorcontrollers::simulation::ElevatorSimSupplier>(
        *m_elevatorSim, [this]() { return m_smc->GetDutyCycle(); }, gearing, circumference,
        period));

    // Build Mechanism2d window.
    double maxH = m_elevatorConfig->GetMaxHeight().value().value();
    double startHValue = startH.value();
    wpi::units::degree_t angle = m_elevatorConfig->GetAngle();
    constexpr double kSoftOffset = 6.0 * 0.0254;  // 6 inches
    constexpr double kHardOffset = 8.0 * 0.0254;  // 8 inches

    m_mechanismWindow.emplace(maxH * 2.0, maxH * 2.0);
    m_mechanismRoot = m_mechanismWindow->GetRoot(m_name + "Root", maxH, 0.0);

    auto smcLowerLimit = m_smc->GetConfig().GetMechanismLowerLimit();
    auto smcUpperLimit = m_smc->GetConfig().GetMechanismUpperLimit();
    if (smcLowerLimit.has_value()) {
      m_mechanismWindow->GetRoot("MinSoft", maxH - kSoftOffset, 0.0)
          ->Append<wpi::MechanismLigament2d>(
              "Limit", m_smc->GetConfig().ConvertFromMechanism(smcLowerLimit.value()).value(),
              angle, 3, wpi::util::Color8Bit{wpi::util::Color::YELLOW});
    }
    if (smcUpperLimit.has_value()) {
      m_mechanismWindow->GetRoot("MaxSoft", maxH - kSoftOffset, 0.0)
          ->Append<wpi::MechanismLigament2d>(
              "Limit", m_smc->GetConfig().ConvertFromMechanism(smcUpperLimit.value()).value(),
              angle, 3, wpi::util::Color8Bit{wpi::util::Color::HOT_PINK});
    }
    m_mechanismWindow->GetRoot("MinHard", maxH - kHardOffset, 0.0)
        ->Append<wpi::MechanismLigament2d>("Limit",
                                           m_elevatorConfig->GetMinHeight().value().value(), angle,
                                           3, wpi::util::Color8Bit{wpi::util::Color::RED});
    m_mechanismWindow->GetRoot("MaxHard", maxH - kHardOffset, 0.0)
        ->Append<wpi::MechanismLigament2d>("Limit",
                                           m_elevatorConfig->GetMaxHeight().value().value(), angle,
                                           3, wpi::util::Color8Bit{wpi::util::Color::LIME_GREEN});

    m_mechanismLigament = m_mechanismRoot->Append<wpi::MechanismLigament2d>(
        m_name, startHValue, angle, 6, m_elevatorConfig->GetSimColor());
    m_setpointLigament = m_mechanismRoot->Append<wpi::MechanismLigament2d>(
        "Setpoint", startHValue, angle, 3, wpi::util::Color8Bit{wpi::util::Color::WHITE});

    wpi::SmartDashboard::PutData(m_name + "/mechanism", &(*m_mechanismWindow));
  }
}

// ---- SmartMechanism overrides -----------------------------------------------

void Elevator::SimIterate() {
  if (m_elevatorSim.has_value() && m_smc->GetSimSupplier()) {
    auto* ss = m_smc->GetSimSupplier();
    ss->UpdateSim();
    m_smc->SimIterate();
    ss->StarveWatchdog();

    if (!m_elevatorConfig->GetMinHeight() || GetHeight() >= *m_elevatorConfig->GetMinHeight()) {
      wpi::sim::RoboRioSim::SetVInVoltage(
          wpi::sim::BatterySim::Calculate({ss->GetCurrentDrawAmps()}));
    }
    VisualizationUpdate();
  }
}

void Elevator::UpdateTelemetry() { m_smc->UpdateTelemetry(); }

void Elevator::VisualizationUpdate() {
  if (m_mechanismLigament) {
    m_mechanismLigament->SetLength(GetHeight().value());
  }
  if (m_setpointLigament) {
    auto setpoint = m_smc->GetMechanismPositionSetpoint();
    if (setpoint.has_value()) {
      m_setpointLigament->SetLength(
          m_smc->GetConfig().ConvertFromMechanism(setpoint.value()).value());
    }
  }
}

std::string Elevator::GetName() const { return m_name; }

// ---- SmartPositionalMechanism overrides -------------------------------------

wpi::cmd::Trigger Elevator::Max() {
  return wpi::cmd::Trigger{[this] {
    return GetHeight() >= m_elevatorConfig->GetMaxHeight().value_or(wpi::units::meter_t{99});
  }};
}

wpi::cmd::Trigger Elevator::Min() {
  return wpi::cmd::Trigger{[this] {
    return GetHeight() <= m_elevatorConfig->GetMinHeight().value_or(wpi::units::meter_t{-99});
  }};
}

// ---- Elevator-specific interface --------------------------------------------

wpi::cmd::CommandPtr Elevator::Run(wpi::units::meter_t height) {
  return wpi::cmd::Run([this, height] { SetMeasurementPositionSetpoint(height); }, {m_subsystem})
      .WithName(m_name + " Run");
}

wpi::cmd::CommandPtr Elevator::Run(std::function<wpi::units::meter_t()> height) {
  return wpi::cmd::Run([this, height] { SetMeasurementPositionSetpoint(height()); }, {m_subsystem})
      .WithName(m_name + " Run Supplier");
}

wpi::cmd::CommandPtr Elevator::RunTo(wpi::units::meter_t height, wpi::units::meter_t tolerance) {
  wpi::cmd::Trigger near = IsNear(height, tolerance).Debounce(wpi::units::second_t{0.1});
  return wpi::cmd::RunOnce([this, height] { SetMeasurementPositionSetpoint(height); },
                            {m_subsystem})
      .AndThen(wpi::cmd::WaitUntil([near] { return near.Get(); }))
      .WithName(m_name + " RunTo");
}

wpi::cmd::CommandPtr Elevator::RunTo(std::function<wpi::units::meter_t()> height, wpi::units::meter_t tolerance) {
  wpi::units::meter_t target = height();
  wpi::cmd::Trigger near = IsNear(target, tolerance).Debounce(wpi::units::second_t{0.1});
  return wpi::cmd::RunOnce([this, target] { SetMeasurementPositionSetpoint(target); },
                            {m_subsystem})
      .AndThen(wpi::cmd::WaitUntil([near] { return near.Get(); }))
      .WithName(m_name + " RunTo Supplier");
}

wpi::cmd::Trigger Elevator::Gte(wpi::units::meter_t height) {
  return wpi::cmd::Trigger{[this, height] { return GetHeight() >= height; }};
}

wpi::cmd::Trigger Elevator::Lte(wpi::units::meter_t height) {
  return wpi::cmd::Trigger{[this, height] { return GetHeight() <= height; }};
}

wpi::cmd::Trigger Elevator::Between(wpi::units::meter_t start, wpi::units::meter_t end) {
  return Gte(start) && (Lte(end));
}

wpi::cmd::Trigger Elevator::IsNear(wpi::units::meter_t height, wpi::units::meter_t within) {
  return wpi::cmd::Trigger{[this, height, within] {
    return std::abs(GetHeight().value() - height.value()) <= within.value();
  }};
}

const config::ElevatorConfig& Elevator::GetConfig() const { return *m_elevatorConfig; }

wpi::math::Translation3d Elevator::GetRelativeMechanismPosition() const {
  if (m_mechanismLigament) {
    return wpi::math::Translation3d{0_m, 0_m, wpi::units::meter_t{m_mechanismLigament->GetLength()}};
  }
  return wpi::math::Translation3d{};
}

void Elevator::SetHeight(wpi::units::meter_t height) { SetMeasurementPositionSetpoint(height); }

wpi::units::meter_t Elevator::GetHeight() const { return m_smc->GetMeasurementPosition(); }

}  // namespace yams::mechanisms::positional

File: yams/java/yams/core/mechanisms/positional/Elevator.java
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

package yams.core.mechanisms.positional;

import static org.wpilib.units.Units.Degrees;
import static org.wpilib.units.Units.Inches;
import static org.wpilib.units.Units.Kilograms;
import static org.wpilib.units.Units.Meters;

import java.util.Optional;
import org.wpilib.framework.RobotBase;
import org.wpilib.math.geometry.Rotation3d;
import org.wpilib.math.geometry.Translation3d;
import org.wpilib.simulation.BatterySim;
import org.wpilib.simulation.ElevatorSim;
import org.wpilib.simulation.RoboRioSim;
import org.wpilib.smartdashboard.Mechanism2d;
import org.wpilib.smartdashboard.MechanismLigament2d;
import org.wpilib.units.measure.Distance;
import org.wpilib.units.measure.LinearVelocity;
import org.wpilib.util.Color;
import org.wpilib.util.Color8Bit;
import yams.core.exceptions.ElevatorConfigurationException;
import yams.core.exceptions.SmartMotorControllerConfigurationException;
import yams.core.mechanisms.config.ElevatorConfig;
import yams.core.mechanisms.config.MechanismPositionConfig;
import yams.core.mechanisms.config.MechanismPositionConfig.Plane;
import yams.core.motorcontrollers.SmartMotorController;
import yams.core.motorcontrollers.SmartMotorControllerConfig;
import yams.core.motorcontrollers.simulation.ElevatorSimSupplier;

/**
 * Elevator mechanism.
 *
 * <p>This core class holds elevator state, physics simulation, and measurement/predicate logic
 * only. Command and Trigger factories live on {@link yams.commands2.mechanisms.Elevator}, which
 * extends this class and has a full usage example in its Javadoc.
 */
public class Elevator extends SmartPositionalMechanism {
  /**
   * Config class for the elevator.
   */
  private final ElevatorConfig m_config;
  /**
   * Simulation for the elevator
   */
  private Optional<ElevatorSim> m_sim = Optional.empty();
  /**
   * Mechanism ligament for the setpoint.
   */
  private MechanismLigament2d m_setpointLigament = null;

  /**
   * Construct the {@link Elevator} class for easy manipulation of an elevator.
   *
   * @param config {@link ElevatorConfig} to set.
   * @param smc    {@link SmartMotorController} to use for the Elevator
   * @implNote Protected so only {@link yams.commands2.mechanisms.Elevator} can construct this.
   */
  protected Elevator(ElevatorConfig config, SmartMotorController smc) {
    m_config = config;
    m_smc = smc;
    SmartMotorControllerConfig smcCfg = smc.getConfig();
    SmartMotorControllerConfig smcConfig = m_smc.getConfig();
    if (config.getTelemetryName().isPresent()) {
      // TODO: Add telemetry units to config.
      m_telemetry.setupTelemetry(getName(), m_smc);
    }

    if (RobotBase.isSimulation()) {
      smc.setupSimulation();
      if (config.getCarriageMass().isEmpty()) {
        throw new ElevatorConfigurationException(
            "Mass is not configured!", "Cannot create simulator", "withCarriageWeight(Mass)");
      }
      if (config.getMinimumHeight().isEmpty()) {
        throw new ElevatorConfigurationException("Minimum height is not configured!",
            "Cannot create simulator", "withHardLimits(Distance,Distance)");
      }
      if (config.getMaximumHeight().isEmpty()) {
        throw new ElevatorConfigurationException("Maximum height is not configured!",
            "Cannot create simulator", "withHardLimits(Distance,Distance)");
      }
      if (smcConfig.getStartingPosition().isEmpty()) {
        throw new SmartMotorControllerConfigurationException("Starting height is not configured!",
            "Cannot create simulator", "withStartingPosition(Distance)");
      }
      if (smcConfig.convertFromMechanism(smcConfig.getStartingPosition().orElseThrow())
              .lt(config.getMinimumHeight().get())
          || smcConfig.convertFromMechanism(smcConfig.getStartingPosition().orElseThrow())
              .gt(config.getMaximumHeight().get())) {
        throw new SmartMotorControllerConfigurationException(
            "Elevator starting height is outside hard limits", "Cannot create simulator",
            "withStartingPosition(Distance)");
      }

      boolean simulateGravity = !config.getIsElevatorHorizontal();

      m_sim = Optional.of(new ElevatorSim(smc.getDCMotor(),
          smcConfig.getGearing().getMechanismToRotorRatio(),
          config.getCarriageMass().get().in(Kilograms),
          smcCfg.getMechanismCircumference().orElseThrow().div(2).div(Math.PI).in(Meters),
          config.getMinimumHeight().get().in(Meters), config.getMaximumHeight().get().in(Meters),
          simulateGravity,
          smcConfig.convertFromMechanism(smcConfig.getStartingPosition().orElseThrow()).in(Meters),
          0.01 / 4096, 0.01 / 4096));
      m_smc.setSimSupplier(new ElevatorSimSupplier(m_sim.get(), m_smc));
      m_mechanismWindow = new Mechanism2d(config.getMaximumHeight().get().in(Meters) * 2,
          config.getMaximumHeight().get().in(Meters) * 2);

      m_mechanismRoot = m_mechanismWindow.getRoot(getName() + "Root",
          config.getMaximumHeight()
              .get()
              .in(Meters) -config.getMechanismPositionConfig()
              .getRelativePosition()
              .orElse(new Translation3d())
              .getX(),
          config.getMechanismPositionConfig()
              .getRelativePosition()
              .orElse(new Translation3d())
              .getZ());

      if (m_smc.getConfig().getMechanismLowerLimit().isPresent()) {
        m_mechanismWindow
            .getRoot("MinSoft",
                config.getMaximumHeight().get().in(Meters)
                    + config.getMechanismPositionConfig()
                        .getRelativePosition()
                        .orElse(new Translation3d())
                        .getX()
                    - Inches.of(6).in(Meters),
                config.getMechanismPositionConfig()
                    .getRelativePosition()
                    .orElse(new Translation3d())
                    .getZ())
            .append(new MechanismLigament2d("Limit",
                m_smc.getConfig()
                    .convertFromMechanism(m_smc.getConfig().getMechanismLowerLimit().get())
                    .in(Meters),
                config.getAngle().in(Degrees), 3, new Color8Bit(Color.YELLOW)));
      }
      if (m_smc.getConfig().getMechanismUpperLimit().isPresent()) {
        m_mechanismWindow
            .getRoot("MaxSoft",
                config.getMaximumHeight().get().in(Meters)
                    + config.getMechanismPositionConfig()
                        .getRelativePosition()
                        .orElse(new Translation3d())
                        .getX()
                    - Inches.of(6).in(Meters),
                config.getMechanismPositionConfig()
                    .getRelativePosition()
                    .orElse(new Translation3d())
                    .getZ())
            .append(new MechanismLigament2d("Limit",
                m_smc.getConfig()
                    .convertFromMechanism(m_smc.getConfig().getMechanismUpperLimit().get())
                    .in(Meters),
                config.getAngle().in(Degrees), 3, new Color8Bit(Color.HOT_PINK)));
      }
      m_mechanismWindow
          .getRoot("MinHard",
              config.getMaximumHeight().get().in(Meters)
                  + config.getMechanismPositionConfig()
                      .getRelativePosition()
                      .orElse(new Translation3d())
                      .getX()
                  - Inches.of(8).in(Meters),
              config.getMechanismPositionConfig()
                  .getRelativePosition()
                  .orElse(new Translation3d())
                  .getZ())
          .append(new MechanismLigament2d("Limit", config.getMinimumHeight().get().in(Meters),
              config.getAngle().in(Degrees), 3, new Color8Bit(Color.RED)));
      m_mechanismWindow
          .getRoot("MaxHard",
              config.getMaximumHeight().get().in(Meters)
                  + config.getMechanismPositionConfig()
                      .getRelativePosition()
                      .orElse(new Translation3d())
                      .getX()
                  - Inches.of(8).in(Meters),
              config.getMechanismPositionConfig()
                  .getRelativePosition()
                  .orElse(new Translation3d())
                  .getZ())
          .append(new MechanismLigament2d("Limit", config.getMaximumHeight().get().in(Meters),
              config.getAngle().in(Degrees), 3, new Color8Bit(Color.LIME_GREEN)));

      m_mechanismLigament = m_mechanismRoot.append(new MechanismLigament2d(getName(),
          smcConfig.convertFromMechanism(smcConfig.getStartingPosition().orElseThrow()).in(Meters),
          config.getAngle().in(Degrees), 6, config.getSimColor()));
      m_setpointLigament = m_mechanismRoot.append(new MechanismLigament2d("Setpoint",
          smcConfig.convertFromMechanism(smcConfig.getStartingPosition().orElseThrow()).in(Meters),
          config.getAngle().in(Degrees), 3, new Color8Bit(Color.WHITE)));
      publishMechanismWindow();
    }
  }

  @Override
  public void updateTelemetry() {
    //    m_telemetry.updatePosition(getHeight());
    //    m_motor.getMechanismPositionSetpoint().ifPresent(m_setpoint ->
    //    m_telemetry.updateSetpoint(m_setpoint));
    m_smc.updateTelemetry();
    m_telemetry.updateLoopTime();
  }

  @Override
  public void simIterate() {
    if (m_sim.isPresent() && m_smc.getSimSupplier().isPresent()) {
      m_smc.getSimSupplier().get().updateSimState();
      m_smc.simIterate();
      m_smc.getSimSupplier().get().starveUpdateSim();
      // It is impossible for an elevator to go below the minimum height, it would break...
      if (m_config.getMinimumHeight().isPresent()
          && getHeight().lt(m_config.getMinimumHeight().get())) {
        //        m_motor.simIterate(RotationsPerSecond.of(0));
        //        m_motor.setEncoderPosition(m_config.getMinimumHeight().get());
      } else {
        RoboRioSim.setVInVoltage(
            BatterySim.calculateDefaultBatteryLoadedVoltage(m_sim.get().getCurrentDraw()));
      }
      visualizationUpdate();
    }
  }

  /**
   * Updates the length of the mechanism ligament to match the current height of the elevator in
   * meters.
   */
  @Override
  public void visualizationUpdate() {
    m_mechanismLigament.setLength(getHeight().in(Meters));
    if (getMotor().getMechanismPositionSetpoint().isPresent()) {
      m_setpointLigament.setLength(m_smc.getConfig()
              .convertFromMechanism(getMotor().getMechanismPositionSetpoint().get())
              .in(Meters));
    }
  }

  /**
   * Get the relative position of the mechanism, taking into account the relative position defined
   * in the
   * {@link MechanismPositionConfig}.
   *
   * @return The relative position of the mechanism as a {@link Translation3d}.
   */
  @Override
  public Translation3d getRelativeMechanismPosition() {
    Plane movementPlane = m_config.getMechanismPositionConfig().getMovementPlane();
    Translation3d mechanismTranslation = new Translation3d(m_mechanismLigament.getLength(),
        new Rotation3d(Plane.YZ == movementPlane ? m_mechanismLigament.getAngle() : 0,
            Plane.XZ == movementPlane ? m_mechanismLigament.getAngle() : 0, 0));
    if (m_config.getMechanismPositionConfig().getRelativePosition().isPresent()) {
      return m_config.getMechanismPositionConfig().getRelativePosition().get().plus(
          mechanismTranslation);
    }
    return mechanismTranslation;
  }

  @Override
  public String getName() {
    return m_config.getTelemetryName().orElse("Elevator");
  }

  /**
   * Get the Height of the Elevator.
   *
   * @return {@link Distance} of the Elevator.
   */
  public Distance getHeight() {
    return m_smc.getMeasurementPosition();
  }

  /**
   * Get the linear velocity of the Elevator.
   *
   * @return {@link LinearVelocity} of the elevator.
   */
  public LinearVelocity getVelocity() {
    return m_smc.getMeasurementVelocity();
  }

  /**
   * Elevator is near a height.
   *
   * @param height {@link Distance} to be near.
   * @param within {@link Distance} within.
   * @return True if the elevator is near another height.
   */
  public boolean isNear(Distance height, Distance within) {
    return getHeight().isNear(height, within);
  }

  @Override
  public boolean isAtMax() {
    if (m_smc.getConfig().getMechanismUpperLimit().isPresent()) {
      return isGte(m_smc.getConfig().convertFromMechanism(
          m_smc.getConfig().getMechanismUpperLimit().get()));
    }
    if (m_config.getMaximumHeight().isPresent()) {
      return isGte(m_config.getMaximumHeight().get());
    }
    throw new ElevatorConfigurationException("Maximum height is not configured!",
        "Cannot create max trigger.", "withHardLimits(Distance,Distance)");
  }

  @Override
  public boolean isAtMin() {
    if (m_smc.getConfig().getMechanismLowerLimit().isPresent()) {
      return isLte(m_smc.getConfig().convertFromMechanism(
          m_smc.getConfig().getMechanismLowerLimit().get()));
    }
    if (m_config.getMinimumHeight().isPresent()) {
      return isLte(m_config.getMinimumHeight().get());
    }
    throw new ElevatorConfigurationException("Minimum height is not configured!",
        "Cannot create min trigger.", "withHardLimits(Distance,Distance)");
  }

  /**
   * Between two heights.
   *
   * @param start Start height.
   * @param end   End height.
   * @return True if the elevator is between the given heights.
   */
  public boolean isBetween(Distance start, Distance end) {
    return isGte(start) && isLte(end);
  }

  /**
   * Less than or equal to height
   *
   * @param height {@link Distance} to check against
   * @return True if the elevator's height is less than or equal to the given height.
   */
  public boolean isLte(Distance height) {
    return getHeight().lte(height);
  }

  /**
   * Greater than or equal to height.
   *
   * @param height Height to check against.
   * @return True if the elevator's height is greater than or equal to the given height.
   */
  public boolean isGte(Distance height) {
    return getHeight().gte(height);
  }

  /**
   * Get the {@link ElevatorConfig} for this {@link Elevator}
   *
   * @return {@link ElevatorConfig} for this {@link Elevator}
   */
  public ElevatorConfig getConfig() {
    return m_config;
  }
}
